### DISCUSSION TITLE
listen to event with minestom event + extra data

### DISCUSSION THREAD

**Discussion by tahmid-23** [Minestom/Minestom#1261] [Help]
I have an `EventNode<Event>`. I would like to listen to specific `PlayerEvent`s. However, these `PlayerEvent`s should be filtered based on a few conditions. One such condition is that I can match the associated `Player` with a `CustomPlayer`, which is my own class.
If a `CustomPlayer` exists, then I would invoke a custom Event class. The event class would look something along the lines of `CustomPlayerEvent<T extends PlayerEvent>(CustomPlayer customPlayer, T event)`. If no such player exists, then this custom event class cannot be constructed and so no event would be fired.

Is there an idiomatic way of doing this? (or a slightly different approach that accomplishes the same result?)

**Comment by FluxCapacitor2** [comment]
I would recommend simply setting the server's player provider, so every player will always be an instance of your own class which extends Player

```java
// Main.java
MinecraftServer.getConnectionManager().setPlayerProvider(CustomPlayer::new)

// CustomPlayer.java
public class CustomPlayer extends Player {
// add custom fields and methods here ...
}
```
Then, every time you have a Player object, it can be cast to CustomPlayer and your custom methods can be used.

**Reply by tahmid-23** [reply]
the point of CustomPlayer is not to act as a Player, I just name it that since it is associated with Players but it is not its own entity
for simplicity, CustomPlayer is just a class for which there is a 1-to-1 correspondence with a Player's UUID
